compiler/resolution/lowerForalls.cpp

```diff
@@ -428,6 +430,45 @@ static void removeVoidReturn(BlockStmt* cloneBody) {
   retexpr->remove();
 }
 
+// Do not create top-level TPVs and such when there are no "top-level" yields
+// i.e. yields outside any parallel constructs.
+// A yield within an 'if' or a serial loop is considered "top-level".
+static std::map<FnSymbol*,bool> toplevelYieldsArePresent;
+typedef std::map<FnSymbol*,bool>::iterator TLVYIterator;
+
+static bool hasToplevelYields(FnSymbol* fn) {
+  std::pair<FnSymbol*,bool> val(fn, false);
+  std::pair<TLVYIterator,bool> result = toplevelYieldsArePresent.insert(val);
+  if (result.second) {
+    // Yes, we inserted a new element. Compute tlvy, store and return it.
+    bool hasTlvy = false;
+    computeHasToplevelYields(fn->body, hasTlvy);
+    result.first->second = hasTlvy;
+    return hasTlvy;
+  } else {
+    // This function is already in the map. Use the cached value.
+    return result.first->second;
+  }
+}
+
+// Call hasToplevelYields() only when we need to know.
+enum HasTLVYTag {
+  TLVY_UNKNOWN,
+  TLVY_ABSENT,
+  TLVY_PRESENT
+};
+
+static bool queryToplevelYields(FnSymbol* fn, HasTLVYTag& tlvy) {
```

Comment:
This function (and the HasTLVYTag argument) appears to be an optimization to avoid repeated map lookups. I doubt that it is necessary or has any significant impact.

Reply:
You are so right. I was overthinking. Will fix.